cmd/zfs/zfs_main.c

```diff
@@ -7244,9 +7260,8 @@ unshare_unmount(int op, int argc, char **argv)
 
 			case OP_MOUNT:
 				if (zfs_unmount(node->un_zhp,
-				    node->un_zhp->zfs_name, flags) != 0)
+				    node->un_mountp, flags) != 0)
```

Comment:
OK.  Let's go ahead and wrap this in `#ifdef __FreeBSD__` as you did initially.  The issue here is that we want to use the dataset name on Linux which is unique while the mountpoint may be ambiguous.